File: Java/src/main/java/org/chess/ruleengine/RuleEngine.java
package org.chess.ruleengine;

import org.chess.domain.GameState;
import org.chess.domain.Move;
import org.chess.domain.RuleViolation;
import org.chess.domain.rules.Rule;
import org.chess.domain.rules.SourcePositionMustNotBeEmpty;

import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Optional;

public class RuleEngine {

    private final List<Rule> rules = List.of(
            new SourcePositionMustNotBeEmpty()
    );

    public MoveResult execute(Move move, GameState gameState) {
        var ruleViolations = applyRules(move, gameState);
        if (!ruleViolations.isEmpty()) {
            return new MoveResult(gameState, ruleViolations);
        }

        var gameStateAfter = movePiece(move, gameState);
        return new MoveResult(gameStateAfter, List.of());
    }

    private Collection<RuleViolation> applyRules(Move move, GameState gameState) {
        var ruleViolations = rules.stream()
                .map(rule -> rule.execute(move, gameState))
                .filter(Optional::isPresent)
                .map(Optional::get);
        return ruleViolations.toList();
    }

    private GameState movePiece(Move move, GameState gameState) {
        var chessPiecePositions = new HashMap<>(gameState.chessPiecePositions());
        var chessPiece = chessPiecePositions.get(move.from());
        chessPiecePositions.remove(move.from());
        chessPiecePositions.put(move.to(), chessPiece);
        return new GameState(chessPiecePositions);
    }
}

File: Kotlin/src/main/kotlin/org/chess/ruleengine/RuleEngine.kt
package org.chess.ruleengine

import org.chess.domain.GameState
import org.chess.domain.Move
import org.chess.domain.PlayerColor
import org.chess.domain.RuleViolation
import org.chess.domain.rules.SourcePositionMustNotBeEmpty
import org.chess.domain.rules.TargetPositionMustNotBeOccupiedByAlly

class RuleEngine {

    private val rules = listOf(
        SourcePositionMustNotBeEmpty(),
        TargetPositionMustNotBeOccupiedByAlly()
    )

    fun execute(move: Move, gameState: GameState): MoveResult {
        val ruleViolations = applyRules(move, gameState)
        if (!ruleViolations.isEmpty()) {
            return MoveResult(gameState, ruleViolations)
        }
        val gameStateAfter = movePiece(move, gameState)
        return MoveResult(gameStateAfter, listOf())
    }

    private fun applyRules(move: Move, gameState: GameState): Collection<RuleViolation> {
        return rules.mapNotNull { it.execute(move, gameState) }
    }

    private fun movePiece(move: Move, gameState: GameState): GameState {
        val chessPiecePositions = gameState.chessPiecePositions.toMutableMap()

        // Assert not null, since a missing chess piece at that position must
        // have already been caught by a rule.
        val chessPiece = chessPiecePositions[move.from]!!

        chessPiecePositions.remove(move.from)
        chessPiecePositions[move.to] = chessPiece

        return GameState(chessPiecePositions, nextPlayer(gameState.currentPlayer))
    }

    private fun nextPlayer(currentPlayer: PlayerColor): PlayerColor {
        return if (currentPlayer === PlayerColor.WHITE) PlayerColor.BLACK else PlayerColor.WHITE
    }
}
